athenahealth/formurlencoder.go

```diff
@@ -0,0 +1,104 @@
+package athenahealth
+
+import (
+	"errors"
+	"io"
+	"net/url"
+	"sort"
+)
+
+const (
+	defaultFormURLEncoderBufferSize = 512
+)
+
+type formURLEncoder struct {
+	entries map[string][]io.Reader
+}
+
+func newFormURLEncoder() *formURLEncoder {
+	return &formURLEncoder{
+		entries: make(map[string][]io.Reader),
+	}
+}
+
+func (f *formURLEncoder) Add(key string, value io.Reader) {
+	f.entries[key] = append(f.entries[key], value)
+}
+
+func (f *formURLEncoder) Encode(w io.Writer) error {
+	keys := make([]string, 0, len(f.entries))
+	for k := range f.entries {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	count := 0
+	for _, key := range keys {
+		for _, reader := range f.entries[key] {
```

Comment:
what if we range f.entries
`for key, reader := range f.entries`

Reply:
We would lose deterministic key ordering. Trying to match the standard lib url.Values encoding as close as possible https://github.com/golang/go/blob/d49017a7c347e57830b87ebbbdbec087bfc1ca92/src/net/url/url.go#L970-L975C21